### ISSUE TITLE
Numeric user id passed to --user interpreted as user name if user name is numeric in container /etc/passwd

### ISSUE THREAD

**Issue by jordmoz** [docker/docker#21436]
**Output of `docker version`:**

```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.10.3
Storage Driver: devicemapper
 Pool Name: docker-8:1-262861-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 543.9 MB
 Data Space Total: 107.4 GB
 Data Space Available: 39.82 GB
 Metadata Space Used: 1.016 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.77 (2012-10-15)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-83-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.86 GiB
Name: jorddev
ID: OL4I:WFOH:3OGA:45VZ:YXHK:A2ME:WYJH:V3WJ:7UOV:AEGY:VJDI:X2JL
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running under VirtualBox through Vagrant.

**Steps to reproduce the issue:**
1.  On docker host: 
```vagrant@jorddev:~$ docker run --rm -it ubuntu:latest /bin/bash```
2.  In the just-started container:
```root@fb5fa3be8269:/# echo '31337:x:0:0:root:/root:/bin/bash' >> /etc/passwd```
3. In another terminal on the same docker host:
```vagrant@jorddev:~$ sudo docker exec -it --user 31337 sad_jones /bin/bash```

**Describe the results you received:**
```
root@fb5fa3be8269:/# id
uid=0(root) gid=0(root) groups=0(root)
```

**Describe the results you expected:**
It should show my user id as 31337, not 0.

**Additional information you deem important (e.g. issue happens only occasionally):**

This could let image creators create malicious images which when run with a specific user id grant root in the container, and thus root to any mounted volumes.

**Comment by thaJeztah** [comment]
What happens if you try this on a regular linux box? i.e.

    echo '31337:x:0:0:root:/root:/bin/bash' >> /etc/passwd

Then

    touch foobar
    chown 31337 foobar

**Comment by jordmoz** [comment]
The behavior is the same as docker:

```vagrant@node0:~$ sudo bash -c 'echo "31337:x:0:0:root:/root:/bin/bash" >> /etc/passwd'
vagrant@node0:~$ touch /tmp/foobar
vagrant@node0:~$ ls -l /tmp/foobar
-rw-rw-r-- 1 vagrant vagrant 0 Mar 24 19:20 /tmp/foobar
vagrant@node0:~$ sudo chown 31337 /tmp/foobar
vagrant@node0:~$ ls -l /tmp/foobar
-rw-rw-r-- 1 root vagrant 0 Mar 24 19:20 /tmp/foobar```

However, even though this is similar to how docker operates, I still need the ability to run a container / exec something in a container with a specific user id, no matter the contents of the container's /etc/passwd.

I would also like to add that adding a user with a numeric username violates the user name regex:

```vagrant@node0:~$ sudo adduser 31338
adduser: Please enter a username matching the regular expression configured
via the NAME_REGEX[_SYSTEM] configuration variable.  Use the `--force-badname'
option to relax this check or reconfigure NAME_REGEX.```

Thus the behavior of `docker run --user=1234` should be to attempt to interpret the flag's argument as a numeric user id first, and if it is not numeric then to interpret it as a alphanumeric username second.

**Comment by thaJeztah** [comment]
I'm not sure we should change that; basically, it follows the way things work in Linux, i.e. request the system if there's a user with that name and if not, treat it as a uid.

copying @NathanMcCauley @diogomonica for input on the security aspect of this

**Comment by crosbymichael** [comment]
Yes, this looks like a bug in the `runc/libcontainer/user` package.  I'll create an issue on runc and we will see about getting this fixed.

**Comment by cyphar** [comment]
This has been fixed by opencontainers/runc#708.

**Comment by thaJeztah** [comment]
Alright, so looks like we need to bump https://github.com/docker/docker/blob/master/hack/vendor.sh#L62
